### ISSUE TITLE
Applying or editing the datadog operator does not trigger rollout upgrade of the underlying agents

### ISSUE THREAD

**Issue by sjanel** [DataDog/datadog-operator#1284]
**Describe what happened:**
Applying (with `kubectl apply -f`) or editing (with `kubectl edit`) the `datadogagent` operator does not rollout update the agents. Even deleting manually the pods does not reload the new `DaemonSet` (for the agent) and `Deployment` for the cluster agent.

**Describe what you expected:**
I expected a rollout update of all agents after any update done on the operator. The only way to make the changes reflected in the daemon set and deploy was to delete them (operator immediately recreates them with the new configuration afterwards). But this method is not very practical as it leads to loss of observability data.

**Steps to reproduce the issue:**
Define an operator as explained in official documentation:
```
apiVersion: datadoghq.com/v2alpha1
kind: DatadogAgent
metadata:
  name: datadog
spec:
  global:
    credentials:
      apiSecret:
        keyName: api-key
        secretName: datadog-secret
  override:
    clusterAgent:
      image:
        name: cluster-agent
        tag: 7.54.0
    nodeAgent:
      image:
        name: agent
        tag: 7.54.0
```

Make an  update (for instance, addition of a label `allowed-metadata: "true"`):
```
apiVersion: datadoghq.com/v2alpha1
kind: DatadogAgent
metadata:
  name: datadog
spec:
  global:
    credentials:
      apiSecret:
        keyName: api-key
        secretName: datadog-secret
  override:
    clusterAgent:
      image:
        name: cluster-agent
        tag: 7.54.0
      labels:
        allowed-metadata: "true"
    nodeAgent:
      image:
        name: agent
        tag: 7.54.0
      labels:
        allowed-metadata: "true"
```

**Additional environment details (Operating System, Cloud provider, etc):**
We are on GKE, deployed the latest datadog operator with helm as detailed in the documentation, we did not test other Kubernetes environments.

**Comment by levan-m** [comment]
Hello, thanks for reaching out. I was unable to reproduce the issue. I used manifests posted above and applied one and later the other in GKE and Kind cluster (with additional `tlsVerify: false`). 

Pods did restart - see the manifest diff below.

```diff
❯ diff pod_before.yaml pod_after.yaml
6c6
<   creationTimestamp: "2024-07-11T16:45:40Z"
---
>   creationTimestamp: "2024-07-11T16:46:08Z"
11a12
>     allowed-metadata: "true"
18,20c19,21
<     controller-revision-hash: c7c7db577
<     pod-template-generation: "1"
<   name: datadog-agent-7vbdl
---
>     controller-revision-hash: cff45989c
>     pod-template-generation: "2"
>   name: datadog-agent-qh27t
29,30c30,31
<   resourceVersion: "149075230"
<   uid: 6b6ccdb2-ba40-4354-b5be-e1304844fa43
---
>   resourceVersion: "149075621"
>   uid: 1a930cc8-168d-4ea0-ab7d-b7afb5c7b0c7
152c153
<       name: kube-api-access-pxdk4
---
>       name: kube-api-access-q8w7j
```

Operator logged following lines timestamp aligning with `creationTimestamp` in `pod_after.yaml`.

```log
{"level":"INFO","ts":"2024-07-11T16:46:07Z","logger":"controllers.DatadogAgent","msg":"Updating Deployment","datadogagent":"datadog/datadog","component":"clusterAgent","deployment.Namespace":"datadog","deployment.Name":"datadog-cluster-agent"}
{"level":"INFO","ts":"2024-07-11T16:46:07Z","logger":"controllers.DatadogAgent","msg":"Creating Deployment","datadogagent":"datadog/datadog","component":"clusterAgent","deployment.Namespace":"datadog","deployment.Name":"datadog-cluster-agent"}
{"level":"INFO","ts":"2024-07-11T16:46:07Z","logger":"controllers.DatadogAgent","msg":"Updating Daemonset","datadogagent":"datadog/datadog","component":"nodeAgent","daemonset.Namespace":"datadog","daemonset.Name":"datadog-agent"}
{"level":"INFO","ts":"2024-07-11T16:46:07Z","logger":"controllers.DatadogAgent","msg":"Creating Daemonset","datadogagent":"datadog/datadog","component":"nodeAgent","daemonset.Namespace":"datadog","daemonset.Name":"datadog-agent"}
```

Could you please repeat the steps and share what's getting logged by Operator.

**Comment by sjanel** [comment]
Hello @levan-m , and thanks for checking this issue.

I checked the logs of the operator after editing the operator by switching a label value from `true` to `false`:

```
{"level":"INFO","ts":"2024-07-20T16:50:18Z","logger":"controllers.DatadogAgent","msg":"Reconciling DatadogAgent","datadogagent":"meta-k8s-vpc/datadog"}
{"level":"INFO","ts":"2024-07-20T16:50:18Z","logger":"controllers.DatadogAgent","msg":"Updating Daemonset","datadogagent":"meta-k8s-vpc/datadog","component":"nodeAgent","daemonset.Namespace":"meta-k8s-vpc","daemonset.Name":"datadog-agent"}
{"level":"INFO","ts":"2024-07-20T16:50:18Z","logger":"controllers.DatadogAgent","msg":"Creating Daemonset","datadogagent":"meta-k8s-vpc/datadog","component":"nodeAgent","daemonset.Namespace":"meta-k8s-vpc","daemonset.Name":"datadog-agent"}
{"level":"INFO","ts":"2024-07-20T16:50:18Z","logger":"controllers.DatadogAgent","msg":"Reconciling DatadogAgent","datadogagent":"meta-k8s-vpc/datadog"}
```

But the change is not applied on my DaemonSet, the label is still set to `true` (and it is not being restarted):

```
allowed-monitoring-rabbitmq: "true"
```

Operator config:

```
spec:
  override:
    clusterAgent:
      ...
    nodeAgent:
      image:
        name: agent
        tag: 7.54.0
      labels:
        ...
        allowed-monitoring-rabbitmq: "false"  # <-- changed this label

```

**Comment by levan-m** [comment]
I tested with Operator 1.7.0. You are probably using earlier version. We did have a bug for the behavior you describe which was fixed with this PR https://github.com/DataDog/datadog-operator/pull/1173 and released in 1.7.0.

I'll close the issue, please reopen if you have any further questions.

**Comment by sjanel** [comment]
Hi @levan-m, indeed upgrading to latest version solved the issue. Thanks a lot!
